### PR TITLE
[QT-506] Use enos scenario samples for testing

### PR DESCRIPTION
**Pull request by ryancragun**
Replace our prior implementation of Enos test groups with the new Enos
sampling feature. With this feature we're able to describe which
scenarios and variant combinations are valid for a given artifact and
allow enos to create a valid sample field (a matrix of all compatible
scenarios) and take an observation (select some to run) for us. This
ensures that every valid scenario and variant combination will
now be a candidate for testing in the pipeline. See QT-504[0] for further
details on the Enos sampling capabilities.

Our prior implementation only tested the amd64 and arm64 zip artifacts,
as well as the Docker container. We now include the following new artifacts
in the test matrix:
* CE Amd64 Debian package
* CE Amd64 RPM package
* CE Arm64 Debian package
* CE Arm64 RPM package

Each artifact includes a sample definition for both pre-merge/post-merge
(build) and release testing.

Changes:
* Remove the hand crafted `enos-run-matrices` ci matrix targets and replace
  them with per-artifact samples.
* Use enos sampling to generate different sample groups on all pull
  requests.
* Update the enos scenario matrices to handle HSM and FIPS packages.
* Simpligy enos scenarios by using shared globals instead of
  cargo-culted locals.

Note: This will require coordination with vault-enterprise to ensure a
smooth migration to the new system. Integrating new scenarios or
modifying existing scenarios/variants should be much smoother after this
initial migration.

[0] https://github.com/hashicorp/enos/pull/102

Signed-off-by: Ryan Cragun <me@ryan.ec>

### CODE REVIEW

File: .github/workflows/enos-run-k8s.yml [hunk 1]
```diff
@@ -42,6 +42,8 @@ jobs:
         uses: hashicorp/action-setup-enos@v1
         with:
           github-token: ${{ secrets.ELEVATED_GITHUB_TOKEN }}
+          # TODO: remove the version pin. This is only use to test a version that is not yet default.
```
Comment (rebwill) [inline .github/workflows/enos-run-k8s.yml:45]: Do we want to remove these before merge, once we're done testing?

File: .github/workflows/test-enos-scenario-ui.yml [hunk 2]
```diff
@@ -74,6 +74,8 @@ jobs:
       - uses: hashicorp/action-setup-enos@v1
         with:
           github-token: ${{ secrets.ELEVATED_GITHUB_TOKEN }}
+          # TODO: remove the version pin. This is only use to test a version that is not yet default.
```
Comment (rebwill) [inline .github/workflows/test-enos-scenario-ui.yml:77]: ^same question as above.

File: .github/workflows/test-run-enos-scenario-matrix.yml [hunk 3]
```diff
@@ -123,32 +115,34 @@ jobs:
       - uses: hashicorp/action-setup-enos@v1
         with:
           github-token: ${{ secrets.ELEVATED_GITHUB_TOKEN }}
+          # TODO: remove the version pin. This is only use to test a version that is not yet default.
```
Comment (rebwill) [inline .github/workflows/test-run-enos-scenario-matrix.yml:119]: ^same question as above

File: enos/enos-globals.hcl [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: BUSL-1.1
+
+globals {
```
Comment (rebwill) [inline enos/enos-globals.hcl:4]: LOVE

File: enos/enos-samples-oss-build.hcl [hunk 5]
```diff
@@ -0,0 +1,142 @@
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: BUSL-1.1
+
+sample "build_oss_linux_amd64_deb" {
```
Comment (rebwill) [inline enos/enos-samples-oss-build.hcl:4]: Soooo cool to finally see this as a real thing getting merged!!!

**Review by rebwill — APPROVED** [review]
Very, very stoked for this 🚀

File: .github/workflows/enos-run-k8s.yml (hunk 1 above)
Comment (ryancragun) [reply .github/workflows/enos-run-k8s.yml:45]: Yeah, I'm testing enos 0.0.20 but haven't released that version to homebrew or to action-setup-enos yet. Once I'm convinced there are no bugs I'll release it in those channels, remove these TODOs and pins and then we'll merge this with the ent equivalent.

File: enos/enos-globals.hcl [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    "ent.fips1402"     = ["ui", "enterprise", "cgo", "hsm", "fips", "fips_140_2", "ent.fips1402"]
+    "ent.hsm"          = ["ui", "enterprise", "cgo", "hsm", "venthsm"]
+    "ent.hsm.fips1402" = ["ui", "enterprise", "cgo", "hsm", "fips", "fips_140_2", "ent.hsm.fips1402"]
+  }
+  distro_version = {
+    "rhel"   = var.rhel_distro_version
+    "ubuntu" = var.ubuntu_distro_version
+  }
+  packages = ["jq"]
+  sample_attributes = {
```
[the remark this answers is not in the document]
Comment (ryancragun) [reply enos/enos-globals.hcl:18]: Yup! In the original design from a million years ago there was a spec for `sample_attributes` you could use for this. I realized that was basically the same thing as the `globals` idea but with a single use. Went with `globals` instead.

**Pull request merged**